app/lib/job/backend.dart

```diff
@@ -0,0 +1,275 @@
+            ..errorCount = current.errorCount + 1
+            ..lastStatus = JobStatus.aborted
+            ..lockedUntil = null
+            ..updatePriority();
+          tx.queueMutations(inserts: [current]);
+          await tx.commit();
+        }
+      });
+    }
+
+    final query = _db.query(Job)
+      ..filter('service =', service)
+      ..filter('state =', JobState.processing)
+      ..filter('lockedUntil <', new DateTime.now().toUtc());
+    await for (Job job in query.run()) {
+      try {
+        if (job.runtimeVersion != versions.runtimeVersion) continue;
+        await _unlock(job);
+      } catch (e, st) {
+        _logger.warning('Unlock of $job failed.', e, st);
+      }
+    }
+  }
+
+  Future checkIdle(JobService service, ShouldProcess shouldProcess) async {
+    Future _schedule(Job job) async {
+      return _db.withTransaction((tx) async {
+        final list = await tx.lookup([job.key]);
+        final Job current = list.single;
+        if (current.state == JobState.idle &&
+            current.lockedUntil == job.lockedUntil) {
+          current
+            ..state = JobState.available
+            ..processingKey = null
+            ..lockedUntil = null;
+          tx.queueMutations(inserts: [current]);
+          await tx.commit();
+        }
+      });
+    }
+
+    Future _extend(Job job) async {
+      return _db.withTransaction((tx) async {
+        final list = await tx.lookup([job.key]);
+        final Job current = list.single;
+        if (current.state == JobState.idle &&
+            current.lockedUntil == job.lockedUntil) {
+          current
+            ..processingKey = null
+            ..lockedUntil = new DateTime.now().toUtc().add(_extendDuration);
+          tx.queueMutations(inserts: [current]);
+          await tx.commit();
+        }
+      });
+    }
+
+    final query = _db.query(Job)
+      ..filter('service =', service)
+      ..filter('state =', JobState.idle)
+      ..filter('lockedUntil <', new DateTime.now().toUtc());
+    await for (Job job in query.run()) {
+      if (job.runtimeVersion != versions.runtimeVersion) continue;
+      try {
+        final process = await shouldProcess(
+            job.packageName, job.packageVersion, job.packageVersionUpdated);
+        if (process) {
+          await _schedule(job);
+        } else {
+          await _extend(job);
+        }
+      } catch (e, st) {
+        _logger.warning('Idle check of $job failed.', e, st);
+      }
+    }
+  }
+
+  Future complete(Job job, JobStatus status, {Duration extendDuration}) async {
+    return _db.withTransaction((tx) async {
+      final items = await tx.lookup([_db.emptyKey.append(Job, id: job.id)]);
+      final Job selected = items.single;
+      if (selected != null && selected.processingKey == job.processingKey) {
+        _logger.info('Updating $job with $status');
+        final errorCount =
+            status == JobStatus.success ? 0 : selected.errorCount + 1;
```

Comment:
Why do we keep jobs that succeeded in the job queue? If I'm reading this right, the Job db will end up containing an entry for each PackageVersion, and we'll keep running through them and extending their lock until we finally decide we need to process them.

Wouldn't it be better to only have jobs that we should process in the queue?

It also looks like we don't remove jobs from old runtime versions?

Reply:
> It also looks like we don't remove jobs from old runtime versions?

When a new service with a newer runtime comes alive, it will rewrite the `Job` entries with the new `runtimeVersion` using `jobMaintenance.syncDatastoreHistory`.

> Wouldn't it be better to only have jobs that we should process in the queue?

I've started with that. The current version gives the benefit of having the last status there, which in theory could be useful, although I can't really point out any immediate use of it. Another potentially useful thing would be to have an "audit trace" in the Datastore, e.g. `Job` would be a parent entity, while a `JobExecution` could contain the main logs and execution time of the given run. Neither is really important or useful for us, shall I remove the entries?

Reply:
One thing that is benefit here: if the successfully generated documentation sets the `lockedUntil` for 90 days, none of the instances need to check whether it needs a new run until it times out. With deleting the entry, it will be checked at least once a day in each running instance.

Reply:
True, let's keep it. I'd missed the update of `runtimeVersion`, so we should be good.

So this is not really a queue, but more of an extension to the PackageVersion table, holding the job statuses for the version. Cool!

Reply:
> So this is not really a queue, but more of an extension to the PackageVersion table, holding the job statuses for the version.

Yeah, that is kind of the role. I am not sure whether the name describes the intent properly. Would you use a different naming for it?

Reply:
`Job` is fine, since it records the jobs and when we need to look at them again. I just had queues in my head.